internal/pkg/deploy/pipeline.go

```diff
@@ -607,14 +634,101 @@ func (stg *PipelineStage) Deployments() ([]DeployAction, error) {
 	return actions, nil
 }
 
-// PreDeployments returns a list of pre-deployment actions for the pipeline stage.
-func (stg *PipelineStage) PreDeployments() ([]Step, error) {
-	return nil, nil
+// PostDeployments returns a list of post-deployment actions for the pipeline stage.
+func (stg *PipelineStage) PostDeployments() ([]PrePostDeployAction, error) {
+	if len(stg.postDeployments) == 0 {
+		return nil, nil
+	}
+
+	var prevActions []orderedRunner
+	if approval := stg.Approval(); approval != nil {
+		prevActions = append(prevActions, approval)
+	}
+	preDeployActions, err := stg.PreDeployments()
+	if err != nil {
+		return nil, err
+	}
+	for i := range preDeployActions {
+		prevActions = append(prevActions, &preDeployActions[i])
+	}
+	deployActions, err := stg.Deployments()
+	if err != nil {
+		return nil, err
+	}
+	for i := range deployActions {
+		prevActions = append(prevActions, &deployActions[i])
+	}
+
+	topo, err := graph.TopologicalOrder(stg.buildPrePostDeploymentsGraph("post"))
+	if err != nil {
+		return nil, fmt.Errorf("find an ordering for deployments: %v", err)
+	}
+
+	var actions []PrePostDeployAction
+	var buildspecPath string
+	for name, conf := range stg.postDeployments {
+		if conf.BuildspecPath != "" {
+			buildspecPath = conf.BuildspecPath
+		}
+		actions = append(actions, PrePostDeployAction{
+			name: name,
+			action: action{
+				prevActions: prevActions,
+			},
+			Build: Build{
+				Image:           defaultPipelineBuildImage,
+				EnvironmentType: defaultPipelineEnvironmentType,
+				BuildspecPath:   buildspecPath,
+			},
+			ranker: topo,
+		})
+	}
+
+	sort.Slice(actions, func(i, j int) bool {
+		return actions[i].Name() < actions[j].Name()
+	})
+	return actions, nil
 }
 
-// PostDeployments returns a list of post-deployment actions for the pipeline stage.
-func (stg *PipelineStage) PostDeployments() ([]Step, error) {
-	return nil, nil
+// Test returns a test for the stage.
+// If the stage does not have any test commands, then returns nil.
+func (stg *PipelineStage) Test() (*TestCommandsAction, error) {
+	if len(stg.testCommands) == 0 {
+		return nil, nil
+	}
+
+	var prevActions []orderedRunner
+	if approval := stg.Approval(); approval != nil {
+		prevActions = append(prevActions, approval)
+	}
+	preDeployActions, err := stg.PreDeployments()
+	if err != nil {
+		return nil, err
+	}
+	for i := range preDeployActions {
+		prevActions = append(prevActions, &preDeployActions[i])
+	}
```

Comment:
same fb as above.

Reply:
pre/dep/post are all independent, so we can't assume any of them

Reply:
So the reason why we need those `prevActions` is because we want to get the `runOrder` which is rank for these nodes in the graph.

If I understand correctly though they are all put into the same stage, the `runOrder` is always from `needApproval` -> `pre` -> `deploy` -> `post`/`test`. And that's why I think we only need to depend on the the previous one to figure out the baseline.

Reply:
but sometimes the previous one doesn't exist 🤔

Reply:
Discussed offline. Will refactor in a separate PR.